github-stats/fetcher/comments_test.go

```diff
@@ -0,0 +1,257 @@
+			[]Comment{
+				{IssueID: 1, CommentUpdatedAt: time.Date(1999, 1, 1, 0, 0, 0, 0, time.UTC)},
+				{IssueID: 1, CommentUpdatedAt: time.Date(2001, 1, 1, 0, 0, 0, 0, time.UTC)},
+				{IssueID: 1, CommentUpdatedAt: time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)},
+				{IssueID: 2, CommentUpdatedAt: time.Date(2002, 1, 1, 0, 0, 0, 0, time.UTC)},
+			},
+			1,
+			time.Date(2001, 1, 1, 0, 0, 0, 0, time.UTC),
+		},
+		// Can pick pull-request comments
+		{
+			[]Comment{
+				{IssueID: 1, PullRequest: true, CommentUpdatedAt: time.Date(1999, 1, 1, 0, 0, 0, 0, time.UTC)},
+				{IssueID: 1, PullRequest: false, CommentUpdatedAt: time.Date(2001, 1, 1, 0, 0, 0, 0, time.UTC)},
+				{IssueID: 1, PullRequest: true, CommentUpdatedAt: time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)},
+			},
+			1,
+			time.Date(2001, 1, 1, 0, 0, 0, 0, time.UTC),
+		},
+		// Can pick issue comments
+		{
+			[]Comment{
+				{IssueID: 1, PullRequest: false, CommentUpdatedAt: time.Date(1999, 1, 1, 0, 0, 0, 0, time.UTC)},
+				{IssueID: 1, PullRequest: true, CommentUpdatedAt: time.Date(2001, 1, 1, 0, 0, 0, 0, time.UTC)},
+				{IssueID: 1, PullRequest: false, CommentUpdatedAt: time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)},
+			},
+			1,
+			time.Date(2001, 1, 1, 0, 0, 0, 0, time.UTC),
+		},
+	}
+
+	for _, test := range tests {
+		db, err := config.CreateDatabase()
+		if err != nil {
+			t.Fatal("Failed to create database:", err)
+		}
+
+		for _, comment := range test.comments {
+			db.Create(&comment)
+		}
+
+		actualLatest := findLatestCommentUpdate(test.issueID, db)
+		if actualLatest != test.expectedLatest {
+			t.Error("Actual:", actualLatest,
+				"doesn't match expected:", test.expectedLatest)
+		}
+	}
+}
+
+func TestUpdateComments(t *testing.T) {
+	config := SQLiteConfig{":memory:"}
+
+	tests := []struct {
+		before           []Comment
+		newIssueComments map[int][]*github.IssueComment
+		newPullComments  map[int][]*github.PullRequestComment
+		after            []Comment
+		updateID         int
+		isPullRequest    bool
+	}{
+		// No new comments
+		{
+			before: []Comment{
+				*makeComment(12, 1, "Body", "Login",
+					time.Date(2000, time.January, 1, 19, 30, 0, 0, time.UTC),
+					time.Date(2001, time.January, 1, 19, 30, 0, 0, time.UTC), true),
+			},
+			newIssueComments: map[int][]*github.IssueComment{},
+			newPullComments:  map[int][]*github.PullRequestComment{},
+			after: []Comment{
+				*makeComment(12, 1, "Body", "Login",
+					time.Date(2000, time.January, 1, 19, 30, 0, 0, time.UTC),
+					time.Date(2001, time.January, 1, 19, 30, 0, 0, time.UTC), true),
+			},
+			updateID:      1,
+			isPullRequest: true,
+		},
+		// New comments, include PR
+		{
+			before: []Comment{
+				*makeComment(12, 1, "Body", "Login",
+					time.Date(2000, time.January, 1, 19, 30, 0, 0, time.UTC),
+					time.Date(2001, time.January, 1, 19, 30, 0, 0, time.UTC), true),
+			},
+			newIssueComments: map[int][]*github.IssueComment{
+				3: {
+					makeGithubIssueComment(2, "IssueBody", "SomeLogin",
```

Comment:
Have you considered storing makeGithubIssueComment()? It might make these easier to read (haven't thought deeply about the side effects)

```
issueBodyComment := makeGithubIssueComment(...)

newIssueComments: map[int][]*github.IssueComment{
  3: { issueBodyComment, anotherBodyComment, etcComment}
}
```